Implement a cache manager for transcodes and other files

pR0Ps opened the pull request:
~This code currently works, but is not ready to merge. Check commit messages for implementation details while it's in a WIP state.~ See below.

@spl0k: Before I dig too deep into this, do you have any general thoughts on how you would like to see something like this implemented?

---------------------------------
Quick summary
-------------
- Adds a Cache class (plus tests for it) that provides an API for managing a cache of files on disk
- Adds two new settings to the configuration file: `cache_size` (default 512MB) and `transcode_cache_size` (default 1GB).
- Creates two cache managers using the settings above: one for general stuff (currently album art and lyrics) and one for transcodes
- Adds the caching of transcoded files to disk for future use
- Modifies the existing image caching to use the cache manager

Longer explanations and justifications
--------------------------------------
The reason I separated out transcodes into an entirely separate cache is that I could imagine a single transcode pushing out a ton of smaller images or other cached content. By separating them it should reduce the number of deletes caused by adding something to the cache.

The cache manager allows for caching a value from a generator via passthrough. This means that a generator can be transparently wrapped to save its output in the cache. The bytes from the generator will be written to a temp file in the cache and yielded back. When it completes, the temp file will be renamed according to the provided cache key. This is how caching transcoded music is implemented.

If multiple generators for the same key are started, they will all write to individual temp files until they complete and race to overwrite each other. Since the key should uniquely represent the content it indexes the files will be identical so overwriting them is harmless.

The cache will store everything for a minimum amount of time (configurable, default is set at 5 minutes). After this time has elapsed, the data can be deleted to free up space. This minimum is so that when you cache a file to the disk you can expect it to be there after, even if another large file is added to the cache and requests that some files are deleted to make space.

To ensure that a file will not be paged out of the cache regardless of the minimum time, there is a `protect` context manager that will refuse the delete the key from the cache as long as it's active.

The cache has a maximum size, however this is more of a recommendation as opposed to a hard limit. The actual size will frequently exceed the limit temporarily until something can be paged out.

spl0k commented:
If more things are being cached (embedded cover art #120 will also use cache) adding some helper methods or class might be a good idea to unify cache handling. Correctly designed it would prevent us from having to bother about path manipulation in random methods. Plus that would give us an entry point to potentially handle cleanup.
Actually rather than managing the cache ourselves we should check (I never did) if there are some libraries that can handle this for us (there must be). Ideally it would allow us to retrieve the cached file path we can then provide to Flask's `send_file()` to benefit from its optimizations rather than maybe having to deal with some random stream.

Other than that I don't have any advice to give you.

pR0Ps commented:
@spl0k Added a new commit, let me know what you think.

spl0k commented:
Looks good. I haven't tested this case but I think this could cache an incomplete file if a client disconnects before the whole file could get transcoded.

Do you have any use case where `protect` could be needed? I see you haven't used it anywhere.

pR0Ps commented:
@spl0k: The `test_cleanup_on_error` test tests the incomplete cache scenario (if an exception is raised during generation the contents up to that point aren't cached).

I added `protect` before I added the minimum time. The idea was that you would manually protect a file for the duration of its use. Something like:
```python
cache_key = "track id"
with cache.protect(cache_key):
    if not cache.has(cache_key):
        with cache.set_fileobj(cache_key) as fp:
            extract_track_image_to_file(some_track, fp)
    return send_file(cache.get(cache_key))
```
That way you wouldn't get into a situation where a resource was bumped out of the cache while you were still using it. With the addition of the `min_time` it's largely redundant but I left it in in case a situation came up where a minimum time wouldn't work (ie. caching something for the duration of the program runtime).

It can definitely be removed if you're worried about complexity/maintainability.

spl0k commented:
I saw the `test_cleanup_on_error` test case, but my concerns are more about the `transcode()` generator that catches exceptions without reraising them afterwards.
https://github.com/spl0k/supysonic/blob/7fa150111299513ede15b0383256f6b9a9105c00/supysonic/api/media.py#L115-L129

I understand the potential need for `protect`, and I feel like the minimum time approach could be safer than this context manager. Flask's `send_file()` can delegate the file transfer to the webserver (with `X-Sendfile`); in that case you could have exited the context while the file is still being sent. This could cause some confusion if it were to happen so I'd rather see this context manager removed.
